### DISCUSSION TITLE
Preprocessor for loop

### DISCUSSION THREAD

**Discussion by mosel00** [dotnet/csharplang#3974] [Language Ideas]
It would be practical to have a code generating for loop. E.g.

    #for INDEX = 0 to 3
        public float x`INDEX`;  
    #endfor 

would generate

    public float x0;
    public float x1;
    public float x2;
    public float x3;

and

    #for INDEX = 0 to 3
        x`INDEX` = a.x`INDEX` + b.x`INDEX`;
    #endfor 

would generate 

    x0 = a.x0 + b.x0;
    x1 = a.x1 + b.x1;
    x2 = a.x2 + b.x2;
    x3 = a.x3 + b.x3;

I don't know what syntax it should have, but it would be nice to have it for e.g. vectors or matrices, where you have a bunch of similar operations, that should be as fast as possible, but multiple types, that only differ in the amount of fields.

**Comment by CyrusNajmabadi** [comment]
How about doing this with source-generators?

**Reply by mosel00** [reply]
I haven't heard of a way to implement source-generators with a few lines of code in the same project.
The last time I've used source-generators, the generated code couldn't be used with Intellisense, because it  consisted only of strings. Furthermore I had to create an additional project for it.

My idea is more of a simple but more elegant way to generate a bit of source code than a complete source generator.

**Reply by CyrusNajmabadi** [reply]
> My idea is more of a simple but more elegant way 

it def feels complex :)  

> The last time I've used source-generators, the generated code couldn't be used with Intellisense, 

it can def be used with intellisense.  Also, we shouldn't be making *new* language features jsut because tooling hasn't caught up with existing alpha language features. :)

**Reply by spydacarnage** [reply]
I didn't think source generators could replace code - only create extra?  How would they work in this case?

**Reply by CyrusNajmabadi** [reply]
Just generate a helper that you would call into :)

**Reply by spydacarnage** [reply]
Hmm, that feels a bit clunky, personally, but then again, I can't see much use for the original request either, so I may not be the target audience here.

The original seems like it just as easily be resolved with arrays and real loops...

**Reply by mosel00** [reply]
It could be resolved with arrays and a normal for loop, but the code would be slower. As a reference: the second example formatted into a normal for loop with an array is about 40% slower.

**Reply by spydacarnage** [reply]
I guess it depends on your scenario.  Would a stackalloc'd array work better?

**Reply by mosel00** [reply]
> it can def be used with intellisense. 

How? Because I could only find source-generation via strings.

> Also, we shouldn't be making _new_ language features jsut because tooling hasn't caught up with existing alpha language features. :)

Is there anything planned to generate source code, that doesn't require a new project?
Also: [here](https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md#proposal) it is said, that code generators can't replace code. So I can't add something like a macro or tag for the generator to look for and replace.

**Reply by mosel00** [reply]
> I guess it depends on your scenario. Would a stackalloc'd array work better?

With stackalloc'd arrays, it is only 11%-17% slower. I think that's coming from the for loop itself. If that's the case, it couldn't get any faster.

**Reply by ufcpp** [reply]
What is really needed is T4 templates implemented as additional files of the source generator and well integrated in dotnet CLI and Visual Studio?

**Comment by theunrepentantgeek** [comment]
This kind of optimization is something I'd expect the compiler and/or JIT runtime to do automatically where appropriate.

I don't think the language needs to be made significantly more complex just to handle an infrastructure scenario that could be handled automatically. 

Having the runtime take care of it would also provide much wider benefit.
